db/db.go

```diff
@@ -79,19 +81,177 @@ const (
+	}
+	return nil
+}
+
+// WritePrices saves the ticker prices to the DB file.
+func (db *Database) WritePrices(ticker string, prices []PriceRow) error {
+	if err := db.createDirs(); err != nil {
+		return errors.Annotate(err, "failed to create DB directories")
+	}
+	if err := writeGob(db.pricesFile(ticker), prices); err != nil {
+		return errors.Annotate(err, "failed to write '%s'", db.pricesFile(ticker))
+	}
+	return nil
+}
+
+// TickerRow for the given ticker. If ticker is not in DB, returns zero value.
+// Tickers are cached in memory upon the first call. Go routine safe.
+func (db *Database) TickerRow(ticker string) (TickerRow, error) {
+	if err := db.cacheTickers(); err != nil {
+		return TickerRow{}, errors.Annotate(err, "failed to load tickers")
+	}
+	r, ok := db.tickers[ticker]
+	if !ok {
+		return TickerRow{}, errors.Reason("no such ticker: %s", ticker)
+	}
+	return r, nil
+}
+
+// Tickers returns the list of tickers satisfying TickerRow data constraints.
+// Tickers are cached in memory upon the first call. Go routine safe.
+func (db *Database) Tickers(c *Constraints) ([]string, error) {
+	if err := db.cacheTickers(); err != nil {
+		return nil, errors.Annotate(err, "failed to load tickers")
+	}
+	tickers := []string{}
+	for t, r := range db.tickers {
+		if c.CheckTicker(t) && c.CheckTickerRow(r) {
+			tickers = append(tickers, t)
+		}
+	}
+	return tickers, nil
+}
+
+// Actions for ticker satisfying the constraints. Actions for all tickers are
+// cached in memory upon the first call. Go routine safe.
+func (db *Database) Actions(ticker string, c *Constraints) ([]ActionRow, error) {
+	if err := db.cacheActions(); err != nil {
+		return nil, errors.Annotate(err, "failed to load actions")
+	}
+	actions, ok := db.actions[ticker]
+	if !ok {
+		return nil, errors.Reason("no actions found for ticker %s", ticker)
+	}
+	res := []ActionRow{}
+	for _, a := range actions {
+		if c.CheckAction(a) {
+			res = append(res, a)
+		}
+	}
+	return res, nil
+}
+
+// Prices for ticker satilfying constraints. Prices are NOT cached in memory,
+// and are read from disk every time. It is probably go routine safe, if the
+// underlying OS allows to open and read the same file multiple times from the
+// same process. Reading different tickers is definitely safe in parallel.
+func (db *Database) Prices(ticker string, c *Constraints) ([]PriceRow, error) {
+	prices := []PriceRow{}
+	if err := readGob(db.pricesFile(ticker), &prices); err != nil {
+		return nil, errors.Annotate(err, "failed to read prices for %s", ticker)
+	}
+	res := []PriceRow{}
+	for _, p := range prices {
+		if c.CheckPrice(p) {
+			res = append(res, p)
+		}
+	}
+	return res, nil
+}
+
+// Resampled rows for a ticker satisfying constraints. All rows for the given
+// frequency are cached in memory upon the first call. Go routine safe.
+func (db *Database) Resampled(ticker string, f ResampledFrequency, c *Constraints) ([]ResampledRow, error) {
+	return nil, nil
+}
```

Comment:
Remove it for now. It's best to implement `Weelky`, `Monthly` and `Quarterly` separately.